### PR TITLE
first attempt to migrate to jakarta

### PR DESCRIPTION
**Pull request by amoscatelli**
@skybber what do you think of this ?

Error messages from the agent disappeared and it kinda works (more or less).

I may need an help double chicking if this is where you wanted to go ...

### CODE REVIEW

**Comment by amoscatelli** [comment]
please notice this can't break anothing legacy, since I created a new collector plugins pom and a specific profile to create the fat jar containing the jakarta plugins

**Comment by amoscatelli** [comment]
#447

**Comment by skybber** [comment]
Good job! Separate plugins for jakarta is better then mix in existing, that could break something as you noticed. Is it possible to make exception and make single Hibernate plugin? I don't see how much it is different from Hibernate plugin. Hibernate is used also at least by Spring folks and they could be confused what plugin to choose.

**Comment by amoscatelli** [comment]
So I strongly suggest keeping separated versions of plugins ...

**Comment by amoscatelli** [comment]
Also @skybber could you please tell me which is the plugin for the EJB instrumentation ?

I can't replace a jakarta.ejb.Stateless and to tell the truth I completely missed a plugin using :

<dependency>
    <groupId>javax.ejb</groupId>
    <artifactId>ejb-api</artifactId>
    <version>3.0</version>
</dependency>

or something similar

**Comment by skybber** [comment]
If it is so, what about creating Hibernate6 plugin? As far as EJBs, I thing we don't support it now, feel free to create ejb plugin if you want. It seems that the next HA version will be little bit bigger then we thought:)

**Comment by amoscatelli** [comment]
You mean you want me to rename the hibernate-jakarta plugin to hibernate6 ?

**Comment by skybber** [comment]
Yes, please.

**Comment by amoscatelli** [comment]
It think it would be misleading too ...
What when hibernate 7 is out, and it still will be a jakarta only compatibile hibernate ?
You may want to upgrade the hibernate6 plugin and then you have to change the name again ...

Wouldn't it be better to rename the current hibernate plugin to hibernate5 (since it is to be unlikely updated) and to rename the hibernate-jakarta plugin to hibernate plugin ?

**Comment by skybber** [comment]
We have already Hibernate3 for Hibernate<=3.0, Hibernate plugin is for 4,5, and Hibernate6 would be for 6+, may be we can ask J.Bubnik what he thinks about this idea.

**Comment by skybber** [comment]
I would not rename existing HIbernate plugin since some people are using HA argument `disablePlugin=...Hibernate,...`

**Comment by amoscatelli** [comment]
Done !
Is this what you wanted ?

**Comment by amoscatelli** [comment]
> I would not recommend renaming the existing HIbernate plugin since some people are using HA argument `disablePlugin=...Hibernate,...`

this is a very good point

**Comment by skybber** [comment]
I would suggest to use Hibernate6, since folks outside of javaEE might not know what jakarta is, but Hibernate6 is clear. We have already Hibernate3. So there are 2 options:

- Hibernate3, Hibernate, JakartaHibernate
- Hibernate3, Hibernate, Hibernate6

From my point of view the second one is better.

**Comment by skybber** [comment]
Also all jakarta plugins should be mutual exclusive with non jakarta plugins for same frameworks. In the runtime, there should not be registered both (jakarta + no-jakarta) running plugins in the core (e.g. RestEasy and JakartaRestEasy). Btw,  you are using the same identification for jakarta plugins as for non-jakarta plugins (RestEasy).

**Comment by amoscatelli** [comment]
They should already be mutual exclusive since they belong to different pom plugins : 

        <module>plugin/hotswap-agent-jakarta-plugins</module> 
        <module>plugin/hotswap-agent-plugins</module>
        
Do you want me to assign unique identification ?
You mean this right ? (for esemple) : 

@Plugin(name = "Hibernate" -> @Plugin(name = "Hibernate6"

**Comment by skybber** [comment]
Yes, unique identification in `@Plugin`, but it is the first step. The second one should be some mechanism guaranteeing that only one of plugin (jakarta/nojakarta) is registered in the code. For instance Weld plugin could be registered from this point from old code:

https://github.com/HotswapProjects/HotswapAgent/blob/master/plugin/hotswap-agent-weld-plugin/src/main/java/org/hotswap/agent/plugin/weld/transformer/BeanDeploymentArchiveTransformer.java#L76

but in your case the transformer for jakarta plugin will register the new plugin as well.

**Comment by amoscatelli** [comment]
@skybber I made the first step.
For the second step I can't really help.
I have a good knowledge of the jakarta transition, but I know nothing of HotswapAgent internal mechanisms.

If you think this is good enough please accept the pull request.
This won't effect anything already in production since the new Jakarta plugins will be excluded from the all-in-one JAR.
Jakarta plugins replace the non jakarta plugins in the all-in-one JAR only if you select the 'jakarta' maven profile during maven build.
Anyway if you have something simple in mind regarding the mutual exclusion mechanism please guide me and I still may be of help.

Thank you in advance !

**Comment by skybber** [comment]
I'll taka a look on it, may be that to keep separate jakarta plugins is not very good idea, let see...

**Comment by amoscatelli** [comment]
@skybber thank you.

Let me know

**Comment by skybber** [comment]
I've pushed CDI  plugin modification. There is a `isJakarta(ClassPool classPool)` method, that is used a jakarta resolver in transformers (in OWB, Weld and Deltaspike plugins). Also there are new OwbJakarta and WeldJakarta plugins. OwbJakarta is passing neaerly all tests, but WeldJakarta does not - last green tests are for v3.x, so could be some breaking changes in v5.0. Could you please take a look at this? For the rest of jakarta-related plugins you can use the `isJakarta(..)` method.

**Comment by amoscatelli** [comment]
@skybber there is something wrong

I think you forgot to commit/push some file.
For example hotswap-agent-weld-jakarta-plugin is missing it's pom file.

In general, the maven project has errors and doesn't even start building.

**Comment by amoscatelli** [comment]
> I've pushed CDI plugin modification. There is a `isJakarta(ClassPool classPool)` method, that is used as a jakarta resolver in transformers (in OWB, Weld and Deltaspike plugins). Also there are new OwbJakarta and WeldJakarta plugins. OwbJakarta is passing nearly all tests, but WeldJakarta does not - last green tests are for v3.x, so could be some breaking changes in v5.0. Could you please take a look at this? For the rest of jakarta-related plugins you can use the `isJakarta(..)` method.

How do I run the tests ?
I merged master in my branch.

You still have one beans.xml based on javax instead of jakarta.

Look at the version from my branch.

I'll try and update the other plugins to use the isJakarta method.

**Comment by amoscatelli** [comment]
@skybber hi again !
I also fixed the class names inside transformers, according to the new package names.
Look at my branch files.

So for example : 

```
        src.append("    Class agC = Class.forName(\"org.hotswap.agent.plugin.weld.command.BeanClassRefreshAgent\", true, this.moduleClassLoaderForBDA);");
```
```
        src.append("    Class agC = Class.forName(\"org.hotswap.agent.plugin.weld_jakarta.command.BeanClassRefreshAgent\", true, this.moduleClassLoaderForBDA);");
```

This could be the reason this wasn't working for you with Weld5.

I am testing this with Wildfly 27, and I have Weld 5 under the hood and everything seems fine.

File: plugin/hotswap-agent-weld-jakarta-plugin/src/main/java/org/hotswap/agent/plugin/weld_jakarta/transformer/BeanDeploymentArchiveTransformer.java [hunk 1]
```diff
@@ -54,7 +54,7 @@ public static void transform(ClassPool classPool, CtClass clazz) throws NotFound
         StringBuilder src = new StringBuilder("{");
         src.append(PluginManagerInvoker.buildInitializePlugin(WeldJakartaPlugin.class));
         src.append(PluginManagerInvoker.buildCallPluginMethod(WeldJakartaPlugin.class, "init"));
-        src.append("org.hotswap.agent.plugin.weld.command.BeanClassRefreshAgent.registerArchive(getClass().getClassLoader(), this, null);");
+        src.append("org.hotswap.agent.plugin.weld_jakarta.command.BeanClassRefreshAgent.registerArchive(getClass().getClassLoader(), this, null);");
```
Comment (amoscatelli) [inline plugin/hotswap-agent-weld-jakarta-plugin/src/main/java/org/hotswap/agent/plugin/weld_jakarta/transformer/BeanDeploymentArchiveTransformer.java:57]: @skybber look at here for example !

**Comment by amoscatelli** [comment]
@skybber I mutual excluded resteasy jakarta/javax versions !

Should I do the same for Hibernate6/Hibernate/Hibernate3 ?

**Pull request closed without merging**